### PR TITLE
Change Triangular to TriDense and add Triangular interface.

### PR DESCRIPTION
**Pull request by btracey**
This brings the Triangular type in line with Dense and Symmetric

### CODE REVIEW

File: mat64/cholesky.go [hunk 1]
```diff
@@ -81,7 +81,7 @@ func (t *Triangular) Cholesky(a *SymDense, upper bool) (ok bool) {
 // SolveCholesky finds the matrix x that solves A * X = B where A = L * L^T or
 // A = U^T * U, and U or L are represented by t. The matrix A must be symmetric
 // and positive definite.
-func (m *Dense) SolveCholesky(t *Triangular, b Matrix) {
+func (m *Dense) SolveCholesky(t *TriDense, b Matrix) {
```
Comment (kortschak) [inline mat64/cholesky.go]: Is there a reason this doesn't also do the RawTriangular dance you have below?

Comment (btracey) [inline mat64/cholesky.go]: If you'd like it can, but the thought was that it would always be used after calling TriDense.Cholesky.

Comment (btracey) [inline mat64/cholesky.go]: But there's no good reason not to allow an arbitrary Triangular.

**Comment by btracey** [comment]
PTAL

File: mat64/cholesky.go (hunk 1 above)
Comment (kortschak) [inline mat64/cholesky.go]: Not fussed really, just wondering.

File: mat64/cholesky.go [hunk 2]
```diff
@@ -107,8 +111,8 @@ func (m *Dense) SolveCholesky(t *Triangular, b Matrix) {
 
 // SolveTri finds the matrix x that solves op(A) * X = B where A is a triangular
 // matrix and op is specified by trans.
-func (m *Dense) SolveTri(a *Triangular, trans bool, b Matrix) {
-	_, n := a.Dims()
+func (m *Dense) SolveTri(a Triangular, trans bool, b Matrix) {
+	n, _ := a.Triangular()
```
Comment (kortschak) [inline mat64/cholesky.go]: Triangle?

File: mat64/triangular.go [hunk 3]
```diff
@@ -6,15 +6,29 @@ import (
...
+// format.
+type TriDense struct {
 	mat blas64.Triangular
 }
 
+type Triangular interface {
+	Matrix
+	// Triangular returns the number of rows/columns in the matrix and if it is
+	// an upper triangular matrix.
+	Triangular() (n int, upper bool)
```
Comment (kortschak) [inline mat64/triangular.go]: Triangle?

**Comment by kortschak** [comment]
It seems to me that a better name for the triangular dims equivalent is Triangle. Triangular is an adjective, to if I say it without its return values, I would guess that it was a badly named IsTriangular() bool. Changing to Triangle() int, bool also has the advantage that the interface name is now different to the interface's method: Triangler.

**Comment by btracey** [comment]
Done

**Comment by kortschak** [comment]
LGTM

**Pull request merged**